src/context.rs

```diff
@@ -36,22 +36,66 @@ impl<N: Name + fmt::Debug> error::Error for UnificationError<N> {
 /// [`Type`]: enum.Type.html
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct Context<N: Name = &'static str> {
-    pub(crate) substitution: HashMap<Variable, Type<N>>,
+    pub(crate) substitution: IndexMap<Variable, Type<N>>,
     next: Variable,
 }
 impl<N: Name> Default for Context<N> {
     fn default() -> Self {
         Context {
-            substitution: HashMap::new(),
+            substitution: IndexMap::new(),
             next: 0,
         }
     }
 }
 impl<N: Name> Context<N> {
     /// The substitution managed by the context.
-    pub fn substitution(&self) -> &HashMap<Variable, Type<N>> {
+    pub fn substitution(&self) -> &IndexMap<Variable, Type<N>> {
         &self.substitution
     }
+    /// The number of constraints in the substitution.
+    pub fn len(&self) -> usize {
+        self.substitution.len()
+    }
+    /// `true` if the substitution has any constraints, else `false`.
+    pub fn is_empty(&self) -> bool {
+        self.substitution.is_empty()
+    }
+    /// Clears the substitution managed by the context.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use polytype::{Type, Context, ptp, tp};
+    /// let mut ctx = Context::default();
+    ///
+    /// // Get a fresh variable
+    /// let t0 = ctx.new_variable();
+    /// let t1 = ctx.new_variable();
+    ///
+    /// let clean = ctx.clone();
+    ///
+    /// if let Type::Variable(N) = t0 {
+    ///     ctx.extend(N, tp![t1]);
+    ///     let dirty = ctx.clone();
+    ///
+    ///     ctx.clean();
+    ///
+    ///     assert_eq!(clean, ctx);
+    ///     assert_ne!(clean, dirty);
+    /// }
+    /// ```
+    pub fn clean(&mut self) {
+        self.substitution = IndexMap::new();
+    }
+    pub fn rollback(&mut self, n: usize) {
```

Comment:
Could you add a docstring? Without looking at the implementation, I wasn’t sure whether this would undo `n` substitutions or maintain the first `n` subs

Reply:
Done - thanks!